File: sdk/src/main/scala/com/horizen/sc2sc/baseprotocol/CrossChainMessageSerializer.scala
package com.horizen.sc2sc.baseprotocol

import com.horizen.account.proposition.AddressPropositionSerializer
import scorex.util.serialization.{Reader, Writer}
import sparkz.core.serialization.SparkzSerializer

object CrossChainMessageSerializer extends SparkzSerializer[CrossChainMessage] {
  override def serialize(s: CrossChainMessage, w: Writer): Unit = {
    w.putInt(s.messageType)
    w.putInt(s.senderSidechain.length)
    w.putBytes(s.senderSidechain)
    w.putInt(s.sender.length)
    w.putBytes(s.sender)
    w.putInt(s.receiverSidechain.length)
    w.putBytes(s.receiverSidechain)
    w.putInt(s.receiver.length)
    w.putBytes(s.receiver)
    w.putInt(s.payload.length)
    w.putBytes(s.payload)
  }

  override def parse(r: Reader): CrossChainMessage = {
    val msgType = r.getInt()
    val senderSidechain = r.getBytes(r.getInt())
    val sender = r.getBytes(r.getInt())
    val receiverSidechain = r.getBytes(r.getInt())
    val receiver =r.getBytes(r.getInt())
    val payload = r.getBytes(r.getInt())
    CrossChainMessage(msgType, senderSidechain, sender, receiverSidechain, receiver, payload)
  }
}

File: sdk/src/main/java/io/horizen/sc2sc/CrossChainMessageSerializer.java
package io.horizen.sc2sc;

import sparkz.core.serialization.SparkzSerializer;
import sparkz.util.serialization.Reader;
import sparkz.util.serialization.Writer;

public class CrossChainMessageSerializer<T extends CrossChainMessage> implements SparkzSerializer<T> {

    private static final CrossChainMessageSerializer serializer;

    static {
        serializer = new CrossChainMessageSerializer<>();
    }

    @Override
    public void serialize(T s, Writer w) {
        w.putShort(s.getProtocolVersion().getVal());
        w.putInt(s.getMessageType());
        w.putBytes(s.getSenderSidechain());
        w.putInt(s.getSender().length);
        w.putBytes(s.getSender());
        w.putBytes(s.getReceiverSidechain());
        w.putInt(s.getReceiver().length);
        w.putBytes(s.getReceiver());
        w.putInt(s.getPayloadHash().length);
        w.putBytes(s.getPayloadHash());
    }

    @Override
    public T parse(Reader reader) {
        short protocolVersion = reader.getShort();
        int messageType = reader.getInt();
        byte[] senderSidechain = reader.getBytes(32);
        byte[] sender = reader.getBytes(reader.getInt());
        byte[] receiverSidechain = reader.getBytes(32);
        byte[] receiver = reader.getBytes(reader.getInt());
        byte[] payload = reader.getBytes(reader.getInt());
        return (T)new CrossChainMessage(
            CrossChainProtocolVersion.fromShort(protocolVersion),
            messageType,
            senderSidechain,
            sender,
            receiverSidechain,
            receiver,
            payload
        );
    }

    public static CrossChainMessageSerializer<CrossChainMessage> getSerializer(){
        return serializer;
    }
}

